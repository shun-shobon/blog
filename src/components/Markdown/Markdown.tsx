import ReactMarkdown from "react-markdown";
import remarkGfm from "remark-gfm";
import SyntaxHighlighter from "react-syntax-highlighter";
import { nord } from "react-syntax-highlighter/dist/cjs/styles/hljs";

export type Props = {
  contents: string;
};

export const Markdown = ({ contents }: Props): JSX.Element => {
  return (
    <section>
      <ReactMarkdown
        remarkPlugins={[remarkGfm]}
        components={{
          h1: ({ node, ...props }) => (
            <h1
              className="font-medium text-3xl border-b border-gray-300 mb-4"
              {...props}
            />
          ),
          h2: ({ node, ...props }) => (
            <h2
              className="font-medium text-2xl border-b border-gray-300 mb-4"
              {...props}
            />
          ),
          h3: ({ node, ...props }) => (
            <h3 className="font-medium text-2xl mb-4" {...props} />
          ),
          h4: ({ node, ...props }) => (
            <h4 className="font-medium text-xl mb-4" {...props} />
          ),
          h5: ({ node, ...props }) => (
            <h5 className="font-medium text-lg mb-4" {...props} />
          ),
          h6: ({ node, ...props }) => (
            <h6 className="text-lg mb-4" {...props} />
          ),
          p: ({ node, ...props }) => (
            <p className="text-base mb-2" {...props} />
          ),
          pre: ({ children }) => (
            <div className="rounded-xl mb-2">{children}</div>
          ),
          blockquote: ({ node, ...props }) => (
            <blockquote
              className="border-l-4 border-gray-200 pl-4 py-2 my-4 text-gray-500"
              {...props}
            />
          ),
          ul: ({ node, ...props }) => (
            <ul className="list-disc pl-5 mb-2" {...props} />
          ),
          ol: ({ node, ...props }) => (
            <ol className="list-decimal pl-5 mb-2" {...props} />
          ),
          strong: ({ node, ...props }) => (
            <strong className="font-bold" {...props} />
          ),
          em: ({ node, ...props }) => <em className="font-italic" {...props} />,
          code: ({ node, inline, className, children, ...props }) => {
            if (inline) {
              return (
                <code
                  className="inline-block rounded bg-gray-100 text-cyan-500 px-1.5"
                  {...props}
                >
                  {children}
                </code>
              );
            }

            const [, language] = /language-(\w+)/.exec(className ?? "") ?? [];
            return (
              <SyntaxHighlighter style={nord} language={language}>
                {children.toString().replace(/\n$/, "")}
              </SyntaxHighlighter>
            );
          },
          del: ({ node, ...props }) => (
            <del className="line-through text-gray-400" {...props} />
          ),
          a: ({ node, ...props }) => (
            <a
              className="text-cyan-500 hover:text-cyan-700 underline"
              {...node.properties}
              {...props}
            />
          ),
        }}
      >
        {contents}
      </ReactMarkdown>
    </section>
  );
};
