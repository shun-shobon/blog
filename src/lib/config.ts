export type Config = {
  isProduction: boolean;
  articlesPath: string;
};

export function getConfig(): Config {
  const isProduction = process.env.NODE_ENV === "production";
  const articlesPath = isProduction ? "articles" : "test/articles";

  return {
    isProduction,
    articlesPath,
  };
}
