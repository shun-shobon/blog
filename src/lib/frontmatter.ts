import * as O from "fp-ts/Option";
import YAML from "yaml";

type Frontmatter = {
  frontmatter: unknown;
  content: string;
};

export function extractFrontmatter(raw: string): O.Option<Frontmatter> {
  const regexp = /^---\n(.+?)\n---\n(.+)/;
  const [, rawFrontmatter, content] = regexp.exec(raw) ?? [];

  if (rawFrontmatter == null || content == null) return O.none;

  const frontmatter: unknown = YAML.parse(rawFrontmatter);
  return O.some({ frontmatter, content });
}
