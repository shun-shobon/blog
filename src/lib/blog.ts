import * as path from "node:path";

import glob from "fast-glob";

import { getConfig } from "./config";

async function findArticles() {
  const { articlesPath } = getConfig();
  const articles = await glob(path.join(articlesPath, "*/main.md"));
}
