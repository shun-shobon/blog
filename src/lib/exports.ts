import * as fs from "node:fs/promises";
import * as path from "node:path";

import { Temporal } from "@js-temporal/polyfill";
import fg from "fast-glob";
import remarkFrontmatter from "remark-frontmatter";
import remarkGfm from "remark-gfm";
import remarkMath from "remark-math";
import remarkParse from "remark-parse";
import { unified } from "unified";

import { ARTICLE_DATABASE_NAME } from "@/config";

import type { ArticleDatabaseSerialized } from "./article";
import type { Article, ArticlePath } from "./plugins";
import {
  MARKDOWN_FILENAME,
  remarkArticle,
  remarkDescriptionList,
  remarkEmbed,
  remarkHeadingId,
  remarkRemovePosition,
  remarkResolveReference,
  remarkSection,
} from "./plugins";

export type ArticleDatabase = {
  articles: Article[];
};

export async function exportArticleDatabase(
  fromDir: string,
  imageExportDir: string,
  dataExportDir: string,
): Promise<void> {
  const articles = await processArticles(fromDir, imageExportDir);
  sortArticles(articles);

  const tags = getAllTags(articles);

  const databaseSerialized: ArticleDatabaseSerialized = {
    tags,
    articles,
  };

  const databasePath = path.join(dataExportDir, ARTICLE_DATABASE_NAME);

  await fs.mkdir(path.dirname(databasePath), { recursive: true });
  await fs.writeFile(databasePath, JSON.stringify(databaseSerialized));
}

async function processArticles(
  fromDir: string,
  toDir: string,
): Promise<Article[]> {
  const slugs = await findArticleSlugs(fromDir);
  const articlePaths = slugs.map((slug) => ({ slug, fromDir, toDir }));

  const articles = await Promise.all(
    articlePaths.map((articlePath) => processArticle(articlePath)),
  );

  return articles;
}

function getAllTags(articles: Article[]): string[] {
  const tagSet = new Set<string>();

  articles.forEach((article) => {
    article.tags.forEach((tag) => tagSet.add(tag));
  });

  return Array.from(tagSet);
}

function sortArticles(articles: Article[]): void {
  articles.sort((a, b) => {
    return Temporal.ZonedDateTime.compare(
      Temporal.ZonedDateTime.from(b.createdAt),
      Temporal.ZonedDateTime.from(a.createdAt),
    );
  });
}

async function findArticleSlugs(basePath: string): Promise<string[]> {
  const pattern = path.join(basePath, "*", MARKDOWN_FILENAME);
  const files = await fg(pattern);

  return files.map((filePath) => {
    return path.basename(path.dirname(filePath));
  });
}

async function processArticle(articlePath: ArticlePath): Promise<Article> {
  const processor = unified()
    .use(remarkParse)
    .use(remarkFrontmatter)
    .use(remarkGfm)
    .use(remarkMath)
    .use(remarkRemovePosition)
    .use(remarkResolveReference)
    .use(remarkEmbed)
    .use(remarkDescriptionList)
    .use(remarkHeadingId)
    .use(remarkSection)
    .use(remarkArticle, articlePath)
    .freeze();

  const filePath = path.join(
    articlePath.fromDir,
    articlePath.slug,
    MARKDOWN_FILENAME,
  );

  const fileContent = await fs.readFile(filePath, "utf-8");

  const article = await processor.run(processor.parse(fileContent));

  return article;
}
