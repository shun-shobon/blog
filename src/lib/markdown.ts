import * as fs from "node:fs/promises";
import * as path from "node:path";

import { Temporal } from "@js-temporal/polyfill";
import fg from "fast-glob";
import remarkFrontmatter from "remark-frontmatter";
import remarkGfm from "remark-gfm";
import remarkMath from "remark-math";
import remarkParse from "remark-parse";
import { unified } from "unified";

import { ARTICLE_SUMMARIES_FILE_NAME } from "@/config";

import type { Article, ArticlePath } from "./plugins";
import {
  MARKDOWN_FILENAME,
  remarkArticle,
  remarkDescriptionList,
  remarkEmbed,
  remarkHeadingId,
  remarkRemovePosition,
  remarkResolveReference,
  remarkSection,
} from "./plugins";

export type ArticleSummaries = {
  summaries: ArticleSummary[];
};

export type ArticleSummary = Pick<
  Article,
  | "title"
  | "plainTitle"
  | "slug"
  | "lead"
  | "createdAt"
  | "updatedAt"
  | "emoji"
  | "tags"
>;

export async function exportArticles(
  fromDir: string,
  imageExportDir: string,
  dataExportDir: string,
): Promise<void> {
  const articles = await processArticles(fromDir, imageExportDir);
  sortArticles(articles);

  await exportArticleSummaries(dataExportDir, articles);

  const promies = articles.map((article) =>
    exportArticle(dataExportDir, article),
  );
  await Promise.all(promies);
}

async function exportArticleSummaries(
  dataExportDir: string,
  articles: Article[],
): Promise<void> {
  const summaries = articles.map(createArticleSummary);
  const articleSummariesPath = path.join(
    dataExportDir,
    ARTICLE_SUMMARIES_FILE_NAME,
  );
  await fs.mkdir(dataExportDir, { recursive: true });
  await fs.writeFile(articleSummariesPath, JSON.stringify({ summaries }));
}

function createArticleSummary(article: Article): ArticleSummary {
  const { title, plainTitle, slug, lead, createdAt, updatedAt, emoji, tags } =
    article;
  return { title, plainTitle, slug, lead, createdAt, updatedAt, emoji, tags };
}

async function exportArticle(
  dataExportDir: string,
  article: Article,
): Promise<void> {
  const articlePath = path.join(dataExportDir, `${article.slug}.json`);
  await fs.writeFile(articlePath, JSON.stringify(article));
}

async function processArticles(
  fromDir: string,
  toDir: string,
): Promise<Article[]> {
  const slugs = await findArticleSlugs(fromDir);
  const articlePaths = slugs.map((slug) => ({ slug, fromDir, toDir }));

  const articles = await Promise.all(
    articlePaths.map((articlePath) => processArticle(articlePath)),
  );

  return articles;
}

function sortArticles(articles: Article[]): void {
  articles.sort((a, b) => {
    const compareCreatedAt = Temporal.PlainDate.compare(
      a.createdAt,
      b.createdAt,
    );

    if (compareCreatedAt !== 0) {
      return -compareCreatedAt;
    }

    if (a.updatedAt && b.updatedAt) {
      return -Temporal.PlainDate.compare(a.updatedAt, b.updatedAt);
    } else if (a.updatedAt && !b.updatedAt) {
      return -1;
    } else if (!a.updatedAt && b.updatedAt) {
      return 1;
    } else {
      return 0;
    }
  });
}

async function findArticleSlugs(basePath: string): Promise<string[]> {
  const pattern = path.join(basePath, "*", MARKDOWN_FILENAME);
  const files = await fg(pattern);

  return files.map((filePath) => {
    return path.basename(path.dirname(filePath));
  });
}

async function processArticle(articlePath: ArticlePath): Promise<Article> {
  const processor = unified()
    .use(remarkParse)
    .use(remarkFrontmatter)
    .use(remarkGfm)
    .use(remarkMath)
    .use(remarkRemovePosition)
    .use(remarkResolveReference)
    .use(remarkEmbed)
    .use(remarkDescriptionList)
    .use(remarkHeadingId)
    .use(remarkSection)
    .use(remarkArticle, articlePath)
    .freeze();

  const filePath = path.join(
    articlePath.fromDir,
    articlePath.slug,
    MARKDOWN_FILENAME,
  );

  const fileContent = await fs.readFile(filePath, "utf-8");

  const article = await processor.run(processor.parse(fileContent));

  return article;
}
