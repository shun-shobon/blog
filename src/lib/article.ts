import type { ArticleSummaries, ArticleSummary } from "@/lib/markdown";
import { ARTICLE_SUMMARIES_FILE_NAME } from "@/lib/markdown";

import type { Article } from "./plugins";

export async function getAllArticleSummaries(): Promise<ArticleSummary[]> {
  return (
    (await import(`data/${ARTICLE_SUMMARIES_FILE_NAME}`)) as ArticleSummaries
  ).summaries;
}

export async function getArticleSummariesByTag(
  tag: string,
): Promise<ArticleSummary[]> {
  const summaries = await getAllArticleSummaries();

  return summaries.filter((summary) => summary.tags.includes(tag));
}

export async function getAllTags(): Promise<string[]> {
  const summaries = await getAllArticleSummaries();

  const tags = new Set<string>();
  summaries.forEach((summary) => summary.tags.forEach((tag) => tags.add(tag)));

  return Array.from(tags);
}

export async function getArticle(slug: string): Promise<Article> {
  return (await import(`data/${slug}.json`)) as Article;
}
