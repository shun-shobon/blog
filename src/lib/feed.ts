import { Temporal } from "@js-temporal/polyfill";
import { Feed } from "feed";

import { ORIGIN, TITLE } from "@/config";
import { getAllArticleSummaries, getAllTags } from "@/lib/article";

import { createOgpImageUrl } from "./ogp-image";
import { getArticleUrl, getUrl } from "./utils";

export async function generateFeed(): Promise<Feed> {
  const summaries = await getAllArticleSummaries();
  const tags = await getAllTags();

  const lastUpdated = summaries[0]?.updatedAt ?? summaries[0]?.createdAt;
  const updated = lastUpdated
    ? new Date(
        Temporal.PlainDateTime.from(lastUpdated)
          .toZonedDateTime("Asia/Tokyo")
          .toInstant().epochMilliseconds,
      )
    : new Date(Temporal.Now.instant().epochMilliseconds);

  const feed = new Feed({
    title: TITLE,
    description:
      "しゅん🌙(@shun_shobon)のブログサイトです。技術記事からポエムまで色々書きます。",
    id: ORIGIN,
    link: ORIGIN,
    language: "ja-JP",
    favicon: getUrl("/favicon.ico").href,
    copyright: "Copyright © 2023 Shuntaro Nishizawa (しゅん🌙)",
    updated,
    author: {
      name: "Shuntaro Nishizawa (しゅん🌙)",
      link: "https://s2n.tech",
    },
  });

  summaries.forEach((summary) => {
    const date = new Date(
      Temporal.PlainDateTime.from(summary.updatedAt ?? summary.createdAt)
        .toZonedDateTime("Asia/Tokyo")
        .toInstant().epochMilliseconds,
    );
    const published = new Date(
      Temporal.PlainDateTime.from(summary.createdAt)
        .toZonedDateTime("Asia/Tokyo")
        .toInstant().epochMilliseconds,
    );

    feed.addItem({
      title: summary.plainTitle,
      id: getArticleUrl(summary.slug).href,
      link: getArticleUrl(summary.slug).href,
      image: {
        url: createOgpImageUrl(summary.plainTitle, summary.tags),
        type: "image/png",
      },
      description: summary.lead,
      category: summary.tags.map((tag) => ({ name: tag })),
      date,
      published,
    });
  });

  tags.forEach((tag) => {
    feed.addCategory(tag);
  });

  return feed;
}
