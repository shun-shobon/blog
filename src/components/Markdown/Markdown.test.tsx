/** @jest-environment jsdom */

import { render } from "@testing-library/react";
import * as E from "fp-ts/Either";

import { Markdown } from "./Markdown";
import { getArticle } from "../../lib/blog";

describe("Markdown", () => {
  it("renders", async () => {
    const article = await getArticle(
      "tests/articles",
      "2021-12-05/test1/main.md",
    );

    expect(E.isRight(article)).toBeTruthy();
    if (E.isLeft(article)) return;

    const contents = article.right.contents;

    const component = render(<Markdown contents={contents} />);

    expect(component.asFragment()).toMatchSnapshot();
  });
});
