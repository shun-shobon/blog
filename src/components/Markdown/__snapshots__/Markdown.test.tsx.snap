// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Markdown renders 1`] = `
<DocumentFragment>
  <h1
    class="font-medium text-3xl border-b border-gray-300 mb-8"
    level="1"
  >
    Test Article 1
  </h1>
  

  <h1
    class="font-medium text-3xl border-b border-gray-300 mb-8"
    level="1"
  >
    Header 1
  </h1>
  

  <h2
    class="font-medium text-2xl border-b border-gray-300 mb-8"
    level="2"
  >
    Header 2
  </h2>
  

  <h3
    class="font-medium text-2xl mb-4"
    level="3"
  >
    Header 3
  </h3>
  

  <h4
    class="font-medium text-xl mb-4"
    level="4"
  >
    Header 4
  </h4>
  

  <h5
    class="font-medium text-lg mb-4"
    level="5"
  >
    Header 5
  </h5>
  

  <h6
    class="text-lg mb-4"
    level="6"
  >
    Header 6
  </h6>
  

  <p
    class="text-base mb-2"
  >
    List
  </p>
  

  <ul
    class="list-disc pl-5 mb-2"
  >
    

    <li>
      item 1
    </li>
    

    <li>
      item 2
    </li>
    

    <li>
      item 3

      <ul
        class="list-disc pl-5 mb-2"
      >
        

        <li>
          Nested item 1
        </li>
        

        <li>
          Nested item 2
        </li>
        

        <li>
          Nested item 3
        </li>
        

      </ul>
      

    </li>
    

  </ul>
  

  <p
    class="text-base mb-2"
  >
    Ordered List
  </p>
  

  <ol
    class="list-decimal pl-5 mb-2"
  >
    

    <li>
      item 1
    </li>
    

    <li>
      item 2
    </li>
    

    <li>
      item 3
    </li>
    

  </ol>
  

  <p
    class="text-base mb-2"
  >
    <strong
      class="font-bold"
    >
      Bold
    </strong>
     
    <em
      class="font-italic"
    >
      Italic
    </em>
     
    <em
      class="font-italic"
    >
      <strong
        class="font-bold"
      >
        Bold Italic
      </strong>
    </em>
     
    <code
      class="inline-block rounded bg-gray-100 text-cyan-500 px-1.5"
    >
      Inline Code
    </code>
     
    <del
      class="line-through text-gray-400"
    >
      Strikethrough
    </del>
  </p>
  

  <div
    class="rounded-xl mb-2"
  >
    <pre
      style="display: block; overflow-x: auto; padding: 0.5em; background: rgb(46, 52, 64); color: rgb(216, 222, 233);"
    >
      <code
        class="language-javascript"
        style="white-space: pre;"
      >
        <span
          style="color: rgb(76, 86, 106);"
        >
          // Code block
        </span>
        <span>
          

        </span>
        <span />
        <span
          style="color: rgb(143, 188, 187);"
        >
          console
        </span>
        <span>
          .log(
        </span>
        <span
          style="color: rgb(163, 190, 140);"
        >
          "Hello World!"
        </span>
        <span>
          );
        </span>
      </code>
    </pre>
  </div>
  

  <blockquote
    class="border-l-4 border-gray-200 pl-4 py-2 my-4 text-gray-500"
  >
    

    <p
      class="text-base mb-2"
    >
      Block Quote
Block Quote
Block Quote
    </p>
    

  </blockquote>
  

  <p
    class="text-base mb-2"
  >
    <a
      class="text-cyan-500 hover:text-cyan-700 underline"
      href="https://www.google.com"
    >
      Link to Google
    </a>
  </p>
</DocumentFragment>
`;
